presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/LeftJoinWithArrayContainsToEquiJoinCondition.java

```diff
@@ -0,0 +1,161 @@
+import static com.facebook.presto.expressions.LogicalRowExpressions.and;
+import static com.facebook.presto.expressions.LogicalRowExpressions.extractConjuncts;
+import static com.facebook.presto.sql.planner.VariablesExtractor.extractAll;
+import static com.facebook.presto.sql.planner.plan.Patterns.join;
+import static com.facebook.presto.sql.relational.Expressions.call;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * When the join condition of a left join has pattern of contains(array, element) where array, we can rewrite it as a equi join condition. For example:
+ * <pre>
+ * - Left Join
+ *      empty join clause
+ *      filter: contains(r_array, l_key)
+ *      - scan l
+ *      - scan r
+ * </pre>
+ * into:
+ * <pre>
+ *     - Left Join
+ *          l_key = field
+ *          - scan l
+ *          - Unnest
+ *              field <- unnest distinct_array
+ *              - project
+ *                  distinct_array := array_distinct(r_array)
+ *                  - scan r
+ *                      r_array
+ * </pre>
+ */
+public class LeftJoinWithArrayContainsToEquiJoinCondition
+        implements Rule<JoinNode>
+{
+    private static final Pattern<JoinNode> PATTERN = join().matching(x -> x.getType().equals(JoinNode.Type.LEFT) && x.getCriteria().isEmpty() && x.getFilter().isPresent());
+    private final FunctionAndTypeManager functionAndTypeManager;
+    private final RowExpressionDeterminismEvaluator determinismEvaluator;
+    private final FunctionResolution functionResolution;
+
+    public LeftJoinWithArrayContainsToEquiJoinCondition(FunctionAndTypeManager functionAndTypeManager)
+    {
+        this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+        this.determinismEvaluator = new RowExpressionDeterminismEvaluator(functionAndTypeManager);
+        this.functionResolution = new FunctionResolution(functionAndTypeManager.getFunctionAndTypeResolver());
+    }
+
+    @Override
+    public Pattern<JoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isRewriteLeftJoinArrayContainsToEquiJoinEnabled(session);
+    }
+
+    @Override
+    public Result apply(JoinNode node, Captures captures, Context context)
+    {
+        RowExpression filterPredicate = node.getFilter().get();
+        List<VariableReferenceExpression> leftInput = node.getLeft().getOutputVariables();
+        List<VariableReferenceExpression> rightInput = node.getRight().getOutputVariables();
+        List<RowExpression> andConjuncts = extractConjuncts(filterPredicate);
+        List<RowExpression> remainingConjuncts = ImmutableList.of();
+        RowExpression array = null;
+        RowExpression element = null;
+        for (RowExpression conjunct : andConjuncts) {
```

Comment:
perhaps you can use Stream.findFirst to do this, something like this:
```
Optional<RowExpression> arrayContains = andConjuncts.stream().findFirst(...)
if (!arrayContains.isPresent()) ...
```

Reply:
I think using findFirst will lead to code like: andConjuncts.stream().filter(...).findFirst()?
The logic in the for loop is not one or two simple functions, I think it may be better to use a for loop here.

Reply:
you're right, it requires a function with a lambda expression to be applied first but I think the function is the same few lines you have below (and can even be refactored into a function):
```
Optional<RowExpression> arrayContains = andConjuncts.stream().filter(conjunct->isSupportedArrayContains(conjunct)).findFirst();
if (!arrayContains.isPresent()) {
    remainingConjuncts = andConjuncts.stream().filter(x -> !x.equals(arrayContains.get()).collect(toImmutableList());
}
```

Reply:
Done